File: Data Structures & Algorithms/count-paths/submission-2.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # curr = [1] * n

        # for i in range(1, m):
        #     for j in range(1, n):
        #         curr[j] += curr[j-1]
        
        # return curr[n-1]
        if m == 1 or n == 1:
            return 1
        
        if m < n:
            return self.uniquePaths(n, m)
        
        res = j = 1
        for i in range(m, m + n - 1):
            res *= i
            res //= j
            j += 1
        return res
        

File: Data Structures & Algorithms/valid-tree/submission-2.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        Set<Integer> visit = new HashSet<>();
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0, -1});
        visit.add(0);

        while (!q.isEmpty()) {
            int[] curr = q.poll();
            int u = curr[0];
            int p = curr[1];
            for (int nbr : adj.get(u)) {
                if (nbr == p) {
                    continue;
                }
                if (visit.contains(nbr)) {
                    return false;
                }
                visit.add(nbr);
                q.offer(new int[]{nbr, u});
            }
        }
        return visit.size() == n;
    }
}
